packages/dds/tree/src/shared-tree/schematizedTree.ts

```diff
@@ -0,0 +1,124 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { unreachableCase } from "@fluidframework/common-utils";
+import {
+	schemaDataIsEmpty,
+	AllowedUpdateType,
+	Compatibility,
+	SimpleObservingDependent,
+} from "../core";
+import {
+	ViewSchema,
+	defaultSchemaPolicy,
+	ViewSchemaCollection,
+	ContextuallyTypedFieldData,
+} from "../feature-libraries";
+import { fail } from "../util";
+import { ISharedTreeBranch } from "./sharedTree";
+
+/**
+ * Takes in a tree and returns a view of it that conforms to the view schema.
+ * The returned view referees to and can edit the provided one: it is not a fork of it.
+ * Updates the stored schema in the tree to match the provided one if requested by config and compatible.
+ *
+ * If the tree is uninitialized (has no nodes or schema at all),
+ * it is initialized to the config's initial tree and the provided schema are stored.
+ * This is done even if `AllowedUpdateType.None`.
+ *
+ * TODO:
+ * - Implement schema-aware API for return type.
+ * - Support adapters for handling out of schema data.
+ * - Support lazy schema updates.
+ * - Improve discoverability of this by either integrating it into the factory or main SharedTree interfaces as a method.
+ * @alpha
+ */
+export function schematizeBranch(
+	tree: ISharedTreeBranch,
+	config: SchematizeConfiguration,
+): ISharedTreeBranch {
+	// Check for empty.
+	// This case could be better handled by making a schematized factory (so it runs only in the "create" case),
+	// but handling it here works well enough, and can be considered a special case of an out of schema handler.
+	// Note that if this becomes a more proper out of schema handler, it should be made lazy.
+	{
+		if (tree.context.root.length === 0 && schemaDataIsEmpty(tree.storedSchema)) {
+			tree.storedSchema.update(config.schema);
+			tree.root = config.initialTree;
+		}
+	}
+
+	// TODO: support adapters and include them here.
+	const viewSchema = new ViewSchema(defaultSchemaPolicy, {}, config.schema);
+	{
+		const compatibility = viewSchema.checkCompatibility(tree.storedSchema);
+		switch (config.allowedSchemaModifications) {
+			case AllowedUpdateType.None: {
+				if (compatibility.read !== Compatibility.Compatible) {
```

Comment:
is there a reason you prefer to write these as "if - fail" rather than asserts?

Reply:
it lets us use the same style for all these checks, some of which do something together than fail inside. Also longer term these checks should do things other than fail.

It doesn't really matter though: they could be asserts and function the same: its just a subjective style choice.